pkg/resource/training_job/hooks.go

```diff
@@ -57,11 +74,94 @@ func (rm *resourceManager) customSetOutput(r *resource) {
 	}
 
 	for _, rule := range r.ko.Status.ProfilerRuleEvaluationStatuses {
+		if ackcompare.IsNotNil(r.ko.Status.ProfilingStatus) && *r.ko.Status.ProfilingStatus == "Disabled" {
+			// Sometimes rule evaluation status will stay in InProgress state.
+			break
+		}
 		if rule.RuleEvaluationStatus != nil && svccommon.IsModifyingStatus(rule.RuleEvaluationStatus, &ruleModifyingStatuses) {
 			svccommon.SetSyncedCondition(r, rule.RuleEvaluationStatus, aws.String("ProfilerRule"), &ruleModifyingStatuses)
 			return
 		}
 	}
 
 	svccommon.SetSyncedCondition(r, trainingJobStatus, &resourceName, &trainingJobModifyingStatuses)
+
+}
+
+// isWarmPoolUpdateable returns a requeue or terminal error depending on the warmpool/training job state
+func (rm *resourceManager) isWarmPoolUpdatable(latest *resource) error {
+	trainingJobStatus := latest.ko.Status.TrainingJobStatus
+	if ackcompare.IsNil(latest.ko.Spec.ResourceConfig.KeepAlivePeriodInSeconds) {
+		return ackerr.NewTerminalError(errors.New("warm pool does not exist"))
+	}
+	if ackcompare.IsNotNil(trainingJobStatus) {
+		if *trainingJobStatus == svcsdk.TrainingJobStatusInProgress {
+			return requeueBeforeUpdate
+		}
+		if *trainingJobStatus == svcsdk.TrainingJobStatusCompleted {
+			if ackcompare.IsNotNil(latest.ko.Status.WarmPoolStatus) {
+				wp_modifying := svccommon.IsModifyingStatus(latest.ko.Status.WarmPoolStatus.Status, &WarmPoolModifyingStatuses)
+				if wp_modifying {
+					return nil
+				} else {
+					return ackerr.NewTerminalError(errors.New("warm pool is in a non updateable state"))
+				}
+			} else {
+				// Sometimes the API (briefly) does not return the WP status even if it completes.
+				// This only occurs for a short time after training job has reached Completed state.
+				return requeueBeforeUpdate
+			}
+		} else {
+			// Training Job is in 'Failed'|'Stopping'|'Stopped' (Terminal)
+			return ackerr.NewTerminalError(errors.New("warm pool is in a non updateable state"))
+		}
+
+	}
+	return nil
+
+}
+
+// customSetOutputUpdateProfiler decides whether the training job is ready/eligible for update
+// depending on the status.
+func (rm *resourceManager) customSetOutputUpdateProfiler(r *resource) error {
+	trainingJobStatus := r.ko.Status.TrainingJobStatus
+	if ackcompare.IsNotNil(trainingJobStatus) {
+		for _, terminalStatus := range TrainingJobTerminalProfiler {
+			if terminalStatus == *trainingJobStatus {
+				return ackerr.NewTerminalError(errors.New("profiler can only be updated when Training Job is in InProgress state"))
+			}
+		}
+	}
+	return nil
+}
+
+// isProfilerRemoved checks if the profiler was removed.
+func (rm *resourceManager) isProfilerRemoved(desired *resource, latest *resource) bool {
+	if ackcompare.IsNil(desired.ko.Spec.ProfilerRuleConfigurations) && ackcompare.IsNotNil(latest.ko.Spec.ProfilerRuleConfigurations) {
+		return true
+	}
+	if ackcompare.IsNil(desired.ko.Spec.ProfilerConfig) && ackcompare.IsNotNil(latest.ko.Spec.ProfilerConfig) {
```

Comment:
I see that in customsetUpdateInput one of the conditions is 1. Customer updates both profiler parameters....
So both ProfilerConfig and ProfilerRuleConfigurations can be updated. However in this code if one of these updated it will return. So in this case the logic for ProfilerConfig won't run? I see that regardless if one or the other is true in these if statements it will run the same handleProfilerRemoved method doesn't seem like that method makes a distinction between the two cases.

I understand i might be missing something, so maybe a comment here would be good to have in the case I am not understanding this correctly.

Reply:
The profiler gets removed when ProfilerConfig or ProfilerRuleConfig (or both) are not present in the spec but were present before. 
>  it will run the same handleProfilerRemoved method doesn't seem like that method makes a distinction between the two cases

Correct, in either case the profiler gets disabled.

>However in this code if one of these updated it will return. So in this case the logic for ProfilerConfig won't run?

It only one gets updated(but not removed from the spec) the this method would return false and customSetInput would eventually get called.

I will document this in a comment